src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs

```diff
@@ -270,11 +270,13 @@ private BoundExpression BindArgListOperator(InvocationExpressionSyntax node, Dia
                             break;
 
                         case BoundKind.ThisReference:
-                            if (InConstructorInitializer && receiver.WasCompilerGenerated)
+                            // Can't call the HasThis method due to EE doing odd things with containing member and its containing type.
+                            if ((InConstructorInitializer || InFieldInitializer) && receiver.WasCompilerGenerated)
```

Comment:
No, it won't get hit. Added tests for that along with a group of other tests (including @AlekseyTs's comments above), will submit PR once I address this and your other comment below.

Reason: For static ctors, this diagnostic gets reported https://github.com/dotnet/roslyn/blob/fd32a7348150294a5f9d3a212da67d41977b2f12/src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs#L484 which then makes this if statement https://github.com/dotnet/roslyn/blob/fd32a7348150294a5f9d3a212da67d41977b2f12/src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs#L546 be false and not call BindDynamicInvocation, which is where the above diff is. The same happens for static fields, but the original diagnostic is reported in a slightly different location a couple lines away https://github.com/dotnet/roslyn/blob/fd32a7348150294a5f9d3a212da67d41977b2f12/src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs#L476